File: 0938/bottom_up_dfs.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <string>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  int rangeSumBST(TreeNode* root, int low, int high) {
    return dfs(root, low, high);
  }

  int dfs(TreeNode* node, int low, int high) {
    if (!node) { return 0; }
    int sum = 0;
    if (low <= node->val && node->val <= high) {
      sum += node->val;
    }
    if (low < node->val) {
      sum += dfs(node->left, low, high);
    }
    if (node->val < high) {
      sum += dfs(node->right, low, high);
    }
    return sum;
  }
};

int main() {
  Solution sol;
  TreeNode* root;
  int low, high;
  int expected;

  root = CreateTreeNode("10 5 15 3 7 # 18");
  low = 7, high = 15;
  expected = 32;
  EXPECT_EQ(expected, sol.rangeSumBST(root, low, high));
  DeleteTreeNode(root);

  root = CreateTreeNode("10 5 15 3 7 13 18 1 # 6");
  low = 6, high = 10;
  expected = 23;
  EXPECT_EQ(expected, sol.rangeSumBST(root, low, high));
  DeleteTreeNode(root);

  exit(EXIT_SUCCESS);
}

File: 0938/bottom_up_dfs.rs
use std::rc::Rc;
use std::cell::RefCell;

use leetcode::{TreeNode, create_treenode};

struct Solution;

impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        return Solution::recursion(&root, low, high);
    }

    fn recursion(root: &Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        let mut sum = 0;
        if root.is_none() {
            return sum;
        }
        let root_ref = root.as_ref().unwrap().borrow();
        if low < root_ref.val {
            sum += Solution::recursion(&root_ref.left, low, high);
        }
        if root_ref.val < high {
            sum += Solution::recursion(&root_ref.right, low, high);
        }
        if low <= root_ref.val && root_ref.val <= high {
            sum += root_ref.val;
        }
        return sum;
    }
}

fn main() {
    /*         10
     *        /  \
     *       5    15
     *      / \     \
     *     3   7     18
     */
    let (root, low, high, expected) = (create_treenode("10 5 15 3 7 # 18"), 7, 15, 32);
    assert_eq!(Solution::range_sum_bst(root, low, high), expected);

    let (root, low, high, expected) = (create_treenode("10 5 15 3 7 13 18 1 # 6"), 6, 10, 23);
    assert_eq!(Solution::range_sum_bst(root, low, high), expected);
}
